### PR TITLE
User DB methods and API Routes

### PR DESCRIPTION
**Pull request by JayHenry07**
Completed db.get_user, db.update_user_activity, db.search, api.login, api.update_contact, and api.search_contacts

### CODE REVIEW

File: db.py [hunk 1]
```diff
@@ -50,42 +53,76 @@ def new_user(self, user: dict) -> tuple:
...
         pass
 
-    def get_user(self, id: int) -> tuple:
+
+    def get_user(self, username: str, password: str) -> tuple:
         """return user and error"""
+        
+        try:
+            self.db.execute('''SELECT * FROM user
+                WHERE username = (?) AND password = (?)''',
```
Comment (VijayStroup) [inline db.py:80]: No need for () around (?)

File: db.py [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 from middleware.hash import hash_password
+from middleware.auth import auth
 
 
 class DB:
     def __init__(self):
         self.con = sqlite3.connect('contact.db', check_same_thread=False)
         self.db = self.con.cursor()
 
+
```
Comment (VijayStroup) [inline db.py:12]: methods, separate from functions, typically only have 1 blank line, not 2.

File: db.py [hunk 3]
```diff
@@ -50,42 +53,76 @@ def new_user(self, user: dict) -> tuple:
...
-    def get_user(self, id: int) -> tuple:
+
+    def get_user(self, username: str, password: str) -> tuple:
         """return user and error"""
+        
+        try:
+            self.db.execute('''SELECT * FROM user
+                WHERE username = (?) AND password = (?)''',
+                username, hash_password(password, username))
+            self.con.commit()
```
Comment (VijayStroup) [inline db.py:82]: do not need to commit with `SELECT` commands

File: db.py [hunk 4]
```diff
@@ -50,42 +53,76 @@ def new_user(self, user: dict) -> tuple:
...
+
+            user = self.db.fetchone()
+            if not user: return (user, None, 401, 'Unauthorized')
+            else: return (user, auth.encode_token(user['id']), 0, None)
```
Comment (VijayStroup) [inline db.py:86]: I don't think the return will be a dict, it returns is a list.
I recommend you make a user dict from the list that you will be returned which relates to the table schema like this:
```py
user = {
    id=user[0],
    record_created=user[1],
...
    password=user[4]
}

auth.encode_token(user['id'])
```

File: db.py [hunk 5]
```diff
@@ -50,42 +53,76 @@ def new_user(self, user: dict) -> tuple:
...
+        except sqlite3.Error as e:
+            return (self.db.fetchone(), None, 500, e)
```
Comment (VijayStroup) [inline db.py:88]: return `None` for the user here too.

File: middleware/auth.py [hunk 6]
```diff
@@ -11,7 +11,8 @@ class Auth:
     def __init__(self):
         self.secret = 'secret'
 
-    def encode_token(self, id: str):
+
```
Comment (VijayStroup) [inline middleware/auth.py:14]: methods have 1 blank line, functions have 2

File: db.py [hunk 7]
```diff
@@ -50,42 +53,76 @@ def new_user(self, user: dict) -> tuple:
...
         string"""
 
-        pass
+        try:
+            self.db.execute('''SELECT * FROM user
+                WHERE id = (?) AND * LIKE %(?)%''', id, search)
+            self.con.commit()
+            return (self.db.fetchall(), 0, None)
+        except sqlite3.Error as e:
+            return (self.db.fetchall(), 500, e)
```
Comment (VijayStroup) [inline db.py:125]: return None when getting an error

File: db.py [hunk 8]
```diff
@@ -50,42 +53,76 @@ def new_user(self, user: dict) -> tuple:
...
+
+    def search(self, id: int, search: str) -> tuple:
+        """return a list of contacts that have a partial match to the search
         string"""
 
-        pass
+        try:
+            self.db.execute('''SELECT * FROM user
+                WHERE id = (?) AND * LIKE %(?)%''', id, search)
+            self.con.commit()
```
Comment (VijayStroup) [inline db.py:122]: no need for commit

**Review by VijayStroup — CHANGES_REQUESTED** [review]
Note: some comments may also be the same thing in other places.

File: db.py (hunk 5 above)
Comment (JayHenry07) [reply db.py:88]: Return None for user on line 85 and 86?

Comment (VijayStroup) [reply db.py:88]: No;
If you find the user return (user, jwt, 0, None)
If you dont return (None, None, code, msg)

File: db.py [hunk 9]
```diff
@@ -32,6 +33,7 @@ def init(self):
 
         self.con.commit()
 
+
```
Comment (VijayStroup) [inline db.py:36]: remove extra line

File: db.py [hunk 10]
```diff
@@ -50,42 +52,83 @@ def new_user(self, user: dict) -> tuple:
                 return (409, 'Username already exists')
             else: return (500, e)
 
+
```
Comment (VijayStroup) [inline db.py:55]: remove extra line

File: db.py [hunk 11]
```diff
@@ -50,42 +52,83 @@ def new_user(self, user: dict) -> tuple:
...
     def new_contact(self, id: int, contact: dict) -> tuple:
         """create a new contact for the user"""
 
         pass
 
+
```
Comment (VijayStroup) [inline db.py:61]: remove extra line

File: db.py [hunk 12]
```diff
@@ -50,42 +52,83 @@ def new_user(self, user: dict) -> tuple:
...
     def get_contacts(self, id: int) -> tuple:
         """return a list of contacts for the user"""
 
         pass
 
-    def get_user(self, id: int) -> tuple:
+
```
Comment (VijayStroup) [inline db.py:73]: remove extra line

File: db.py (hunk 7 above)
Comment (VijayStroup) [reply db.py:131]: still need to return None when we except the error

File: middleware/auth.py [hunk 13]
```diff
@@ -23,6 +23,7 @@ def encode_token(self, id: str):
             algorithm='HS256'
         )
 
+
```
Comment (VijayStroup) [inline middleware/auth.py:26]: remove extra line

File: middleware/auth.py [hunk 14]
```diff
@@ -39,6 +40,7 @@ def decode_token(self, token):
 
         return user
 
+
```
Comment (VijayStroup) [inline middleware/auth.py:43]: remove extra line

**Review by VijayStroup — CHANGES_REQUESTED** [review]

File: db.py [hunk 15]
```diff
@@ -60,32 +61,69 @@ def del_contact(self, id: int, contact: dict) -> tuple:
 
         pass
 
+
```
Comment (VijayStroup) [inline db.py:64]: remove extra line

File: db.py [hunk 16]
```diff
@@ -60,32 +61,69 @@ def del_contact(self, id: int, contact: dict) -> tuple:
...
         pass
 
-    def get_user(self, id: int) -> tuple:
+    def get_user(self, username: str, password: str) -> tuple:
         """return user and error"""
+        
+        try:
+            self.db.execute('''SELECT * FROM user
+                WHERE username = ? AND password = ?''',
+                username, hash_password(password, username))
```
Comment (VijayStroup) [inline db.py:76]: bind statements need to be tuples so change the variables you change in to `(username, hash_password(password, username))`

File: db.py [hunk 17]
```diff
@@ -60,32 +61,69 @@ def del_contact(self, id: int, contact: dict) -> tuple:
...
-        pass
+        time = str(datetime.utcnow())
+
+        try:
+            self.db.execute('''UPDATE user
+                SET last_logged_in = ? WHERE id = ?''', time, id)
+            self.con.commit()
+
+            self.db.execute('SELECT * FROM user WHERE id = ?', id)
+            return (self.db.fetchone(), 0, None)
```
Comment (VijayStroup) [inline db.py:105]: We don't care about getting user info in this method, you can just return a error and message for both parts

File: db.py [hunk 18]
```diff
@@ -60,32 +61,69 @@ def del_contact(self, id: int, contact: dict) -> tuple:
...
 
     def update_contact(self, id: int, contact_id: int, contact: dict) -> tuple:
         """update contact where id is user id and contact_id is the contact's
         id with the new contact dict payload"""
 
         pass
 
-    def search(self, id: int, name: str) -> tuple:
-        """return a list of contacts that have a partial match to the name
+
```
Comment (VijayStroup) [inline db.py:116]: remove extra line

File: db.py [hunk 19]
```diff
@@ -60,32 +61,69 @@ def del_contact(self, id: int, contact: dict) -> tuple:
...
+    def search(self, id: int, search: str) -> tuple:
+        """return a list of contacts that have a partial match to the search
         string"""
 
-        pass
+        try:
+            self.db.execute('''SELECT * FROM user
+                WHERE id = ? AND * LIKE %?%''', id, search)
```
Comment (VijayStroup) [inline db.py:123]: bind statement needs to be an iterable: (id, search)

**Review by VijayStroup — CHANGES_REQUESTED** [review]

File: db.py [hunk 20]
```diff
@@ -65,27 +66,60 @@ def get_contacts(self, id: int) -> tuple:
...
 
-    def search(self, id: int, name: str) -> tuple:
-        """return a list of contacts that have a partial match to the name
+    def search(self, id: int, search: str) -> tuple:
+        """return a list of contacts that have a partial match to the search
         string"""
 
-        pass
+        try:
+            self.db.execute('''SELECT * FROM user
```
Comment (VijayStroup) [inline db.py:118]: We are actually searching from the contacts table not the user table but you are using the foreign key to only search contacts associated with that user.

**Review by VijayStroup — CHANGES_REQUESTED** [review]

**Review by VijayStroup — APPROVED** [review]

**Pull request merged**